<head>
<meta http-equiv="Content-Type" content="text/html;  charset=utf-8">

<style type="text/css">

body {
 margin-top: 5%; 
 margin-bottom: 5%; 
 margin-right: 10%; 
 margin-left: 10%; 
 background-color: #000000; 
 color: #109080; 
 font-size: 24px; 
}
a { text-decoration: none; 
	color: #28B8B8; }
a:visited {	color: #389898; }
A:hover {	color: yellow; }
A:focus {	color: red; }
code { color: pink;  background-color: #302030}
pre { color: gray;  font-size: 16px; background-color: #001010}
div {display: inline-block;  width: 48%;  padding: 2px;  border-radius: 4px;  
border: 1px solid gray;  margin: 3px;  vertical-align:middle; }
</STYLE>
</head>
<body>

<h1>Manipulate Biological Data Using Biostrings Package Exercises(Part 3)</h1>

Bioinformatics is an amalgamation Biology and Computer science.<br/>
Biological Data is manipulated using Computers and Computer softwareâ€™s in Bioinformatics.<br/> Biological Data includes DNA; RNA &amp; Proteins. <br/>
DNA &amp; RNA is made of Nucleotide which are our genetic material in which we are coded.<br/>
Our Structure and Functions are done by protein, which are build of Amino acids<br/>
<br/>
In this exercise we try compare between DNAs, RNAs &amp; Amino Acid Sequences to fund out the relationships.<br/>
Comparison is done using sequence alignment or sequence comparison techniques.<br/><br/>
There are two types of sequence alignment exists.<br/><br/>
1.Pairwise alignment<br/>
2.Multiple Sequence Alignment</p>
<p>Pairwise alignment refers to comparison between two sequences, where as Multiple Sequence Alignment refers to comparing more than two sequences.</p>
<p>In the exercises below we cover how we can do pairwise alignment using Biostrings package in Bioconductor.</p>
<p><b>Install Packages Biostrings</b></p>
<p>Answers to the exercises are available <a href="http://r-exercises.com/2017/06/06/manipulate-biological-data-using-biostrings-package-solutionspart-3/">here</a>.</p>

<p><strong>Exercise 1</strong></p>
<p>Create two DNA strings and do pairwise alignment using local, global and overlap alignment techniques and print the score.</p>

<pre>
####################
#                  #
#    Exercise 1    #
#                  #
####################

library(Biostrings)
DNA1 &lt;-DNAString(&quot;ATAGTAGATGCGGCGCGCTAGAG&quot;)
DNA2 &lt;-DNAString(&quot;ATGTCGTCGTAGTCGTGTGTCAC&quot;)
matrix &lt;- nucleotideSubstitutionMatrix(match = 1, mismatch = -3, baseOnly = TRUE)
gAlign &lt;-pairwiseAlignment(DNA1, DNA2, substitutionMatrix = matrix,
                    gapOpening = 3, gapExtension = 1)
lAlign &lt;-pairwiseAlignment(DNA1, DNA2, type = &quot;local&quot;, substitutionMatrix = matrix,
                    gapOpening = 3, gapExtension = 1)
oAlign &lt;-pairwiseAlignment(DNA1, DNA2, type = &quot;overlap&quot;, substitutionMatrix = matrix,
                    gapOpening = 3, gapExtension = 1)
print(lAlign)
</pre>

<pre>## Local PairwiseAlignmentsSingleSubject (1 of 1)
## pattern: [4] GTAG 
## subject: [9] GTAG 
## score: 4
</pre>

<pre>print(gAlign)
</pre>

<pre>## Global PairwiseAlignmentsSingleSubject (1 of 1)
## pattern: [1] ATAGTAGATGCGGCGCGCTAG---------AG 
## subject: [1] AT-GT-----CGTCG---TAGTCGTGTGTCAC 
## score: -24
</pre>

<pre>print(oAlign)
</pre>

<pre>## Overlap PairwiseAlignmentsSingleSubject (1 of 1)
## pattern: [1] &quot;&quot;
## subject: [1] &quot;&quot;
## score: 0
</pre>


<p><strong>Exercise 2</strong></p>
<p>Create two DNA strings and do pairwise alignment and write the alignment to an .aln file.</p>

<pre>####################
#                  #
#    Exercise 2    #
#                  #
####################

library(Biostrings)
DNA1 &lt;-DNAString(&quot;ATAGTAGATGCGGCGCGCTAGAG&quot;)
DNA2 &lt;-DNAString(&quot;ATGTCGTCGTAGTCGTGTGTCAC&quot;)
matrix &lt;- nucleotideSubstitutionMatrix(match = 1, mismatch = -3)
gAlign &lt;-pairwiseAlignment(DNA1, DNA2, substitutionMatrix = matrix,
                           gapOpening = 3, gapExtension = 1)
print(&quot;Written to File, Available in your default R directory&quot;)
</pre>

<pre>## [1] &quot;Written to File, Available in your default R directory&quot;
</pre>

<pre>writePairwiseAlignments(gAlign,file=&quot;Alignment.aln&quot;)
</pre>

<p><strong>Exercise 3</strong></p>
<p>Create two Amino acid strings and do pairwise alignment </p>

<pre>
####################
#                  #
#    Exercise 3    #
#                  #
####################

library(Biostrings)
Amino1 &lt;- AAString(&quot;MALWMRLLPAAAFLALWGPD&quot;)
Amino2 &lt;- AAString(&quot;MALWMRLLPLLALLALWGPD&quot;)
pAlign&lt;-pairwiseAlignment(Amino1,Amino2)
print(pAlign)
</pre>

<pre>## Global PairwiseAlignmentsSingleSubject (1 of 1)
## pattern: [1] MALWMRLLPAAAFLALWGPD 
## subject: [1] MALWMRLLPLLALLALWGPD 
## score: 54.44525
</pre>


<p><strong>Exercise 4</strong></p>
<p>Create two Amino acid strings and do pairwise alignment using BLOSUM62 substitution matrix.</p>

<pre>####################
#                  #
#    Exercise 4    #
#                  #
####################

library(Biostrings)
Amino1 &lt;- AAString(&quot;MALWMRLLPAAAFLALWGPD&quot;)
Amino2 &lt;- AAString(&quot;MALWMRLLPLLALLALWGPD&quot;)
pAlign&lt;-pairwiseAlignment(Amino1,Amino2,substitutionMatrix = &quot;BLOSUM62&quot;, gapOpening = 3, gapExtension = 1)
print(pAlign)
</pre>

<pre>## Global PairwiseAlignmentsSingleSubject (1 of 1)
## pattern: [1] MALWMRLLPAAAFLALWGPD 
## subject: [1] MALWMRLLPLLALLALWGPD 
## score: 93
</pre>



<p><strong>Exercise 5</strong> </p>
<p>Create two Amino acid strings and do pairwise alignment using BLOSUM100 substitution matrix</p>

<pre>####################
#                  #
#    Exercise 5    #
#                  #
####################

library(Biostrings)
Amino1 &lt;- AAString(&quot;MALWMRLLPAAAFLALWGPD&quot;)
Amino2 &lt;- AAString(&quot;MALWMRLLPLLALLALWGPD&quot;)
pAlign&lt;-pairwiseAlignment(Amino1,Amino2,substitutionMatrix = &quot;BLOSUM100&quot;, gapOpening = 3, gapExtension = 1)
print(pAlign)
</pre>

<pre>## Global PairwiseAlignmentsSingleSubject (1 of 1)
## pattern: [1] MALWMRLLPAAAFLALWGPD 
## subject: [1] MALWMRLLPLLALLALWGPD 
## score: 167
</pre>


<p><strong>Exercise 6</strong></p>
<p>Create two Amino acid strings and do pairwise alignment using PAM250 substitution matrix</p>

<pre>####################
#                  #
#    Exercise 6    #
#                  #
####################

library(Biostrings)
Amino1 &lt;- AAString(&quot;MALWMRLLPAAAFLALWGPD&quot;)
Amino2 &lt;- AAString(&quot;MALWMRLLPLLALLALWGPD&quot;)
pAlign&lt;-pairwiseAlignment(Amino1,Amino2,substitutionMatrix = &quot;PAM250&quot;, gapOpening = 3, gapExtension = 1)
print(pAlign)
</pre>

<pre>## Global PairwiseAlignmentsSingleSubject (1 of 1)
## pattern: [1] MALWMRLLPAAAFLALWGPD 
## subject: [1] MALWMRLLPLLALLALWGPD 
## score: 107
</pre>


<p><strong>Exercise 7</strong></p>
<p>Compare between BLOSUM62 substitution matrix of R and that of the NCBI Database using any two amino acid of your choice.</p>

<pre>####################
#                  #
#    Exercise 7    #
#                  #
####################

library(Biostrings)
data(BLOSUM62)
rvalues&lt;-BLOSUM62[&quot;B&quot;, &quot;D&quot;]
ncbifile &lt;- &quot;ftp://ftp.ncbi.nih.gov/blast/matrices/BLOSUM62&quot;
blosum &lt;- as.matrix(read.table(ncbifile))
ncbivalues&lt;-blosum[&quot;B&quot;, &quot;D&quot;]
print(rvalues)
</pre>

<pre>## [1] 4
</pre>

<pre>print(ncbivalues)
</pre>

<pre>## [1] 4
</pre>


<p><strong>Exercise 8</strong></p>
<p>Do pairwise alignment using Needlemann Wunch Alignment algorithm and print the score, suppress any warnings.</p>

<pre>####################
#                  #
#    Exercise 8    #
#                  #
####################

options(warn=-1)
library(Biostrings)
Amino1 &lt;- AAString(&quot;MALWMRLLPAAAFLALWGPD&quot;)
Amino2 &lt;- AAString(&quot;MALWMRLLPLLALLALWGPD&quot;)
align&lt;-needwunsQS(Amino1,Amino2, substmat = &quot;BLOSUM62&quot;)
print(align)
</pre>

<pre>## Global PairwiseAlignments (1 of 1)
## pattern: [1] MALWMRLLPAAAFLALWGPD 
## subject: [1] MALWMRLLPLLALLALWGPD 
## score: 93
</pre>

<p><strong>Exercise 9</strong></p>

<p>Create two DNA Strings and translate the same to amino acids and do pairwise alignment between the amino acid sequences</p>

<pre>####################
#                  #
#    Exercise 9    #
#                  #
####################

library(Biostrings)
DNA1 &lt;-DNAString(&quot;ATAGTAGATGCGGCGCGCTAGAG&quot;)
DNA2 &lt;-DNAString(&quot;ATGTCGTCGTAGTCGTGTGTCAC&quot;)
translatedAmino1&lt;-translate(DNA1)
translatedAmino2&lt;-translate(DNA2)
align&lt;-pairwiseAlignment(translatedAmino1,translatedAmino2)
print(align)
</pre>

<pre>## Global PairwiseAlignmentsSingleSubject (1 of 1)
## pattern: [1] IVDAAR* 
## subject: [1] MSS*SCV 
## score: -43.67328
</pre>

<p><strong>Exercise 10</strong></p>
<p>Create two RNA Strings and translate the same to amino acids and do pairwise alignment between the amino acid sequences</p>

<pre>####################
#                  #
#    Exercise 10    #
#                  #
####################

library(Biostrings)
RNA1 &lt;-RNAString(&quot;AUAGUAGAUGCGGCGCGCUAGAG&quot;)
RNA2 &lt;-RNAString(&quot;AUGUCGUCGUAGUCGUGUGUCAC&quot;)
translatedRNA1&lt;-translate(RNA1)
translatedRNA2&lt;-translate(RNA2)
align&lt;-pairwiseAlignment(translatedRNA1,translatedRNA2)
print(align)
</pre>

<pre>## Global PairwiseAlignmentsSingleSubject (1 of 1)
## pattern: [1] IVDAAR* 
## subject: [1] MSS*SCV 
## score: -43.67328
</pre>




</body></html>
